### PR TITLE
Support Scenarios in SolaraViz visualization

### PR DESCRIPTION
**Pull request by falloficaruss**
### Summary
This PR adds first-class support for `Scenarios` in `SolaraViz`. It allows users to visualize and interact with models that utilize the new Mesa `Scenario` system, ensuring that parameters are correctly routed to either the Model or the Scenario during instantiation and resets.

### Motive
With the introduction of `Scenarios` in Mesa (#3103, #3168), there was a disconnect in the visualization layer. 
`SolaraViz` previously tried to pass all user-defined sliders and parameters directly into `Model.__init__`. This caused a `TypeError` when a model used a Scenario, because parameters intended for the Scenario were rejected by the Model's initialization function. This PR solves the routing problem and provides a clear UI separation between Model and Scenario parameters.

### Implementation
The implementation follows the "Option 1" (Explicit API) and "Option 2" (Auto-detection) suggestions from the issue:

- `SolaraViz` Parameter Routing: Enhanced `SolaraViz` to automatically introspect the model's `scenario` attribute and intelligently split parameters between Model and Scenario based on their respective signatures.
- Scenario Reconstruction on Reset: Updated `ModelController` and `SimulatorController`. When the "Reset" button is clicked, it now captures the current values of scenario sliders, reconstructs the specific `Scenario` subclass instance, and passes it into the `Model` constructor via the `scenario` keyword argument.
- UI Enhancements: Updated `ModelCreator` to render Scenario-bound sliders in a dedicated "Scenario Parameters" section beneath the main Model parameters, providing visual clarity on parameter ownership.
- Tests: Added comprehensive test coverage in both `tests/visualization/test_solara_viz.py` (testing core functionality like parameter splitting logic, `ModelCreator` with scenarios, and `SolaraViz` integration) and `tests/visualization/test_solara_viz_scenarios.py` (testing scenario-specific edge cases and type hints).

### Usage Examples
Auto Detection:
If your sliders match field names in your `Scenario` class, `SolaraViz` handles the routing automatically:

```python
class MyScenario(Scenario):
    density: float = 0.7

class MyModel(Model):
    def __init__(self, height=40, width=40, scenario=None):
        super().__init__(scenario=scenario)

# 'density' is auto-detected as a Scenario parameter
page = SolaraViz(
    MyModel(),
    model_params={
        "height": 40,
        "density": Slider("Density", 0.7, 0, 1, 0.1)
    }
)
```

### Additional Notes
This change is fully backward compatible. Models not using Scenarios will function exactly as before.
Fixes: #3176

### CODE REVIEW

**Comment by EwoutH** [comment]
Thanks for the PR. However, I feel we didn’t completere finish the discussion in https://github.com/mesa/mesa/issues/3176 on how we would integrate scenario variables exactly. Let’s get full consensus on the design first.

API design is important, since once we commit to it, we have to support it for a long time.

See also https://github.com/mesa/mesa/blob/main/CONTRIBUTING.md#mesa-development-process

**Comment by quaquel** [comment]
I suggest simplifying this by only solving the current scenario problem. That is, keep the reflection stuff to figure out what is in the scenario object and what is on the model init, and use that for the ModelController. Leave the slider stuff untouched.

**Comment by falloficaruss** [comment]
@quaquel Any thoughts here?

**Comment by quaquel** [comment]
Please include test code in the existing test files. @Sahil-Chhoker, I would appreciate your input here as well. 

Basically, for now, I just want to ensure that solara-viz is compatible with scenarios.

File: mesa/visualization/solara_viz.py [hunk 1]
```diff
@@ -535,7 +537,35 @@ def do_reset():
             10,
             f"creating new {model.value.__class__} instance with {model_parameters.value}",
         )
-        model.value = model.value = model.value.__class__(**model_parameters.value)
+        kwargs = {}
+        scenario_kwargs = {}
+
+        # Check if the model has a scenario and split parameters accordingly
+        if hasattr(model.value, "scenario") and model.value.scenario:
```
Comment (quaquel) [inline mesa/visualization/solara_viz.py]:
```suggestion
        if getattr(model.value, "scenario", None):
```

File: mesa/visualization/solara_viz.py [hunk 2]
```diff
@@ -535,7 +537,35 @@ def do_reset():
...
+            scenario_class = model.value.scenario.__class__
```
Comment (quaquel) [inline mesa/visualization/solara_viz.py]:
```suggestion
            scenario_class = type(model.value.scenario)
```

**Comment by quaquel** [comment]
I tested it with #3167. It does not yet work correctly because rng is not handled properly. 

My suggestion is that you port Boltzmann to use a new scenario, modeled on #3167, and use that to have a proper integration test.

File: mesa/examples/advanced/epstein_civil_violence/agents.py
Comment (quaquel) [inline mesa/examples/advanced/epstein_civil_violence/agents.py:1]: Why did you add in Epstein instead of e.g.,. Boltzmann? this now classes with #3176.

Comment (falloficaruss) [reply mesa/examples/advanced/epstein_civil_violence/agents.py:1]: I applied the scenario changes to Epstein as a quick integration target.
I’ll move the scenario‑based integration work to Boltzmann (modeled on #3167) and adjust the tests accordingly.

**Comment by quaquel** [comment]
It won't run for me locally: 
```
raise ValueError(
ValueError: Mesa's visualization requires the use of keyword arguments to ensure the parameters are passed to Solara correctly. Please ensure all model parameters are of form param=value
```

File: mesa/examples/basic/boltzmann_wealth_model/model.py [hunk 3]
```diff
@@ -27,19 +36,42 @@ class BoltzmannWealth(Model):
...
-            rng (int, optional): Random rng. Defaults to None.
+            scenario: BoltzmannScenario object containing model parameters.
         """
-        super().__init__(rng=rng)
+        if args:
+            if len(args) > 3:
+                raise ValueError("Expected at most 3 positional args: n, width, height")
+            arg_keys = ("n", "width", "height")
+            for key, value in zip(arg_keys, args):
+                kwargs.setdefault(key, value)
```
Comment (quaquel) [inline mesa/examples/basic/boltzmann_wealth_model/model.py]: why, please just keep the example simple and as close as possible to the original version.e

File: mesa/examples/basic/boltzmann_wealth_model/model.py [hunk 4]
```diff
@@ -27,19 +36,42 @@ class BoltzmannWealth(Model):
...
+            }
+            scenario = BoltzmannScenario(rng=rng, **scenario_kwargs)
+        else:
+            scenario_kwargs = {
+                key: kwargs.pop(key)
+                for key in ("n", "width", "height")
+                if key in kwargs
+            }
+            for key, value in scenario_kwargs.items():
+                setattr(scenario, key, value)
```
Comment (quaquel) [inline mesa/examples/basic/boltzmann_wealth_model/model.py]: please just use 

```python
self.my_param = scenario.param
```

Also, remove the old kwargs stuff. in the if clause. It should work only with the scenario object, and if that is None, you instantiate a Scenario object with its default values.

File: mesa/visualization/solara_viz.py [hunk 5]
```diff
@@ -533,9 +535,45 @@ def do_reset():
...
+                if k in scenario_defaults and k not in model_init_params:
+                    scenario_kwargs[k] = v
+                else:
+                    kwargs[k] = v
+
+            # Handle seeding continuity and ensure scenario instantiation
+            if "rng" in model_parameters.value:
+                scenario_kwargs["rng"] = model_parameters.value["rng"]
+            elif "seed" in model_parameters.value:
+                scenario_kwargs["rng"] = model_parameters.value["seed"]
```
Comment (quaquel) [inline mesa/visualization/solara_viz.py]: in my view, scenario contains rng not the other way around.

also seed is deprecated anyway.

**Comment by falloficaruss** [comment]
> It won't run for me locally:
> 
> ```
> raise ValueError(
> ValueError: Mesa's visualization requires the use of keyword arguments to ensure the parameters are passed to Solara correctly. Please ensure all model parameters are of form param=value
> ```

Can u check it again?

File: mesa/examples/basic/boltzmann_wealth_model/model.py [hunk 6]
```diff
@@ -27,19 +36,22 @@ class BoltzmannWealth(Model):
         datacollector (DataCollector): Collects and stores model data
     """
 
-    def __init__(self, n=100, width=10, height=10, rng=None):
+    def __init__(self, scenario=None, rng=None):
```
Comment (quaquel) [inline mesa/examples/basic/boltzmann_wealth_model/model.py]:
```suggestion
    def __init__(self, scenario=None):
```

rng is contained in scenario so not needed anymore in the signature

File: mesa/examples/basic/boltzmann_wealth_model/model.py [hunk 7]
```diff
@@ -27,19 +36,22 @@ class BoltzmannWealth(Model):
...
             rng (int, optional): Random rng. Defaults to None.
+            scenario: BoltzmannScenario object containing model parameters.
         """
-        super().__init__(rng=rng)
+        if scenario is None:
+            scenario = BoltzmannScenario(rng=rng)
 
-        self.num_agents = n
-        self.grid = OrthogonalMooreGrid((width, height), random=self.random)
+        super().__init__(scenario=scenario, rng=rng)
```
Comment (quaquel) [inline mesa/examples/basic/boltzmann_wealth_model/model.py]:
```suggestion
        super().__init__(scenario=scenario)
```

File: mesa/examples/basic/boltzmann_wealth_model/model.py [hunk 8]
```diff
@@ -27,19 +36,22 @@ class BoltzmannWealth(Model):
...
         Args:
-            n (int, optional): Number of agents. Defaults to 100.
-            width (int, optional): Grid width. Defaults to 10.
-            height (int, optional): Grid height. Defaults to 10.
             rng (int, optional): Random rng. Defaults to None.
+            scenario: BoltzmannScenario object containing model parameters.
         """
-        super().__init__(rng=rng)
+        if scenario is None:
+            scenario = BoltzmannScenario(rng=rng)
```
Comment (quaquel) [inline mesa/examples/basic/boltzmann_wealth_model/model.py]:
```suggestion
            scenario = BoltzmannScenario()
```

File: mesa/examples/basic/boltzmann_wealth_model/model.py [hunk 9]
```diff
@@ -11,6 +11,15 @@
...
 from mesa.examples.basic.boltzmann_wealth_model.agents import MoneyAgent
+from mesa.experimental.scenarios import Scenario
+
+
+class BoltzmannScenario(Scenario):
+    """Scenario parameters for the Boltzmann Wealth model."""
+
+    n: int = 100
+    width: int = 10
+    height: int = 10
```
Comment (quaquel) [inline mesa/examples/basic/boltzmann_wealth_model/model.py:22]:
```suggestion
    height: int = 10
    rng: SeedLike|None = None
```

File: mesa/examples/basic/boltzmann_wealth_model/st_app.py [hunk 10]
```diff
@@ -19,7 +19,9 @@
 )
 height = st.slider("Select Grid Height", min_value=10, max_value=100, step=10, value=15)
 width = st.slider("Select Grid Width", min_value=10, max_value=100, step=10, value=20)
-model = BoltzmannWealth(num_agents, height, width)
+model = BoltzmannWealth(
+    scenario=BoltzmannScenario(n=num_agents, width=width, height=height)
```
Comment (quaquel) [inline mesa/examples/basic/boltzmann_wealth_model/st_app.py]:
```suggestion
    scenario=BoltzmannScenario(n=num_agents, width=width, height=height, rng=42)
```

**Comment by falloficaruss** [comment]
> Member

Applied the changes but I think this PR needs further fixing.

**Comment by falloficaruss** [comment]
- Aligned the Boltzmann example with the scenario-first design. `BoltzmannWealth` now only takes scenario (no rng in the model signature). If scenario is `None`, it instantiates `BoltzmannScenario()` with defaults and uses parameters via `scenario.*`

- Updated all call sites/tests to pass `rng` through `BoltzmannScenario(rng=...)` instead of `BoltzmannWealth(rng=...)`. This keeps RNG ownership in the scenario as requested.

- Also removed `SolaraViz’s` override that injected `rng/seed` into scenarios from model `params` to avoid reversing the ownership.

This keeps the example close to the original while making the scenario the single source of truth for parameters (including RNG).

**Comment by EwoutH** [comment]
Thanks for working on this.

I think the goal of scenarios (in the long term) is also to decomplicate handling model parameters, like in the visualisation. Right now it seems it mainly complicates it, at least on the source code side. Any reflections on how this will develop in the future?

**Comment by quaquel** [comment]
In my opinion, the coupling between the UI and models is too tight and that is causing a lot of the problems.

**Comment by falloficaruss** [comment]
I agree the current integration adds complexity, especially in the viz layer. The intent is that this is a transitional cost.

As dicussed in https://github.com/mesa/mesa/discussions/2972 in the longer term the goal is to make the UI only deal with a single surface: `scenario`. When models consistently expose parameters via scenarios, the visualization doesn’t need to inspect model signatures or split params manually. It just binds controls to scenario fields and recreates the model with a scenario instance. That should reduce coupling between UI and model internals.

Right now we’re in a hybrid state (models with and without scenarios), so the UI has to support both and it looks more complex than the steady state. Once scenarios are the default pattern, we can remove the fallback logic and the visualization code should get simpler.

File: mesa/examples/basic/boltzmann_wealth_model/app.py [hunk 11]
```diff
@@ -34,6 +32,11 @@ def agent_portrayal(agent):
     },
     "width": 10,
     "height": 10,
+    "rng": {
+        "type": "InputText",
+        "value": 42,
+        "label": "Random Seed",
+    },
```
Comment (quaquel) [inline mesa/examples/basic/boltzmann_wealth_model/app.py]: can you put this back at the top. It is is misleading in the diff.

File: mesa/visualization/solara_viz.py [hunk 12]
```diff
@@ -205,7 +204,9 @@ def set_reactive_use_threads(value):
                 )
         with solara.Card("Model Parameters"):
             ModelCreator(
-                model, model_params, model_parameters=reactive_model_parameters
+                model,
+                model_params,
+                model_parameters=reactive_model_parameters,
```
Comment (quaquel) [inline mesa/visualization/solara_viz.py]: these are also not really changes. Please check the diff and revert anything that is not a real change (unless this is due to ruff or so, but I doubt it)

File: mesa/visualization/solara_viz.py [hunk 13]
```diff
@@ -469,6 +470,7 @@ def ModelController(
     if model_parameters is None:
         model_parameters = {}
     model_parameters = solara.use_reactive(model_parameters)
+
```
Comment (quaquel) [inline mesa/visualization/solara_viz.py]: same here

File: mesa/visualization/solara_viz.py [hunk 14]
```diff
@@ -533,9 +535,41 @@ def do_reset():
...
+        # Check if the model has a scenario and split parameters accordingly
+        if getattr(model.value, "scenario", None):
+            scenario_class = type(model.value.scenario)
+            # We assume scenario parameters are those in scenario class but NOT in model init
+            # Check for _scenario_defaults first (Mesa's Scenario mechanism) or fallback to signature
+            scenario_defaults = getattr(model.value.scenario, "_scenario_defaults", {})
+            if not scenario_defaults:
+                # Fallback to signature inspection if _scenario_defaults is missing
+                scenario_defaults = inspect.signature(scenario_class).parameters
+
```
Comment (quaquel) [inline mesa/visualization/solara_viz.py]: I am not sure we need this. Just assume or check if the scenario is a subclass of mesa's Scenario and raise a ValueError otherwise.

File: mesa/visualization/solara_viz.py [hunk 15]
```diff
@@ -533,9 +535,41 @@ def do_reset():
...
+            model_init_params = inspect.signature(type(model.value).__init__).parameters
+
+            for k, v in model_parameters.value.items():
+                if k in scenario_defaults and k not in model_init_params:
+                    scenario_kwargs[k] = v
```
Comment (quaquel) [inline mesa/visualization/solara_viz.py]: why this check?

File: mesa/visualization/solara_viz.py [hunk 16]
```diff
@@ -660,9 +696,33 @@ def do_reset():
         simulator.reset()
         visualization_pause_event.clear()
         pause_step_event.clear()
-        model.value = model.value = model.value.__class__(
-            simulator=simulator, **model_parameters.value
-        )
+
```
Comment (quaquel) [inline mesa/visualization/solara_viz.py:720]: how much of the code below is redundant with the code above and if so can it be handled in a separate function to avoid code duplication.

File: mesa/visualization/solara_viz.py [hunk 17]
```diff
@@ -766,7 +826,7 @@ def ModelCreator(
     model_parameters = solara.use_reactive(model_parameters)
 
     solara.use_effect(
-        lambda: _check_model_params(model.value.__class__.__init__, user_params),
+        lambda: _check_model_params(type(model.value).__init__, user_params),
```
Comment (quaquel) [inline mesa/visualization/solara_viz.py]: please remove this. You can put in a separate PR for just the `__class__` to `type` changes and we can quickly merge that because it is a good idea.

File: mesa/visualization/solara_viz.py (hunk 15 above)
Comment (falloficaruss) [reply mesa/visualization/solara_viz.py]: This check is to avoid ambiguous routing in hybrid models.
We only send a key to `scenario_kwargs` when it is scenario-defined and not explicitly declared in model `__init__`.
If a name exists in both places, we currently treat it as a model kwarg to preserve backward compatibility.
If we want stricter semantics, I can switch to raising on overlap (or always prefer scenario) and remove this compatibility behavior.

**Comment by quaquel** [comment]
I still get the error shown below when I try to run this locally. 

```
Traceback (most recent call last):
  File "/opt/anaconda3/lib/python3.13/site-packages/reacton/core.py", line 1900, in _reconsolidate
    effect()
    ~~~~~~^^
  File "/opt/anaconda3/lib/python3.13/site-packages/reacton/core.py", line 1131, in __call__
    self._cleanup = self.callable()
                    ~~~~~~~~~~~~~^^
  File "/Users/user_name/Documents/GitHub/mesa/mesa/visualization/solara_viz.py", line 831, in <lambda>
    lambda: _check_model_params(type(model.value).__init__, user_params),
            ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/user_name/Documents/GitHub/mesa/mesa/visualization/solara_viz.py", line 887, in _check_model_params
    raise ValueError(f"Invalid model parameter: {name}")
ValueError: Invalid model parameter: rng
```

**Comment by falloficaruss** [comment]
> I still get the error shown below when I try to run this locally.
> 
> ```
> Traceback (most recent call last):
>   File "/opt/anaconda3/lib/python3.13/site-packages/reacton/core.py", line 1900, in _reconsolidate
>     effect()
>     ~~~~~~^^
>   File "/opt/anaconda3/lib/python3.13/site-packages/reacton/core.py", line 1131, in __call__
>     self._cleanup = self.callable()
>                     ~~~~~~~~~~~~~^^
>   File "/Users/user_name/Documents/GitHub/mesa/mesa/visualization/solara_viz.py", line 831, in <lambda>
>     lambda: _check_model_params(type(model.value).__init__, user_params),
>             ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/Users/user_name/Documents/GitHub/mesa/mesa/visualization/solara_viz.py", line 887, in _check_model_params
>     raise ValueError(f"Invalid model parameter: {name}")
> ValueError: Invalid model parameter: rng
> ```

Will you take a look again?

**Comment by quaquel** [comment]
I now get `asyncio.exceptions.CancelledError`. Does it run locally for you, and, if so, can you elaborate on how you are testing it?

**Comment by falloficaruss** [comment]
> I now get `asyncio.exceptions.CancelledError`. Does it run locally for you, and, if so, can you elaborate on how you are testing it?

I am currently testing the visualization using `pytest` combined with `solara.render`.
In Solara and Reacton, `asyncio.CancelledError` should actually be normal and expected behavior. When a user clicks "Reset" or "Pause," Solara cancels the previous simulation task and (if resetting) starts a new one. I'm not totally sure about this and I would like to get some help on this.
For now I will explicitly catch and handle these silently.

**Comment by quaquel** [comment]
I had a closer look at this and I know what is going on. First, revert your last commit. Errors should never be past over in silence. Also, it is not the root cause of the problem

My problem now happens when I hit reset. This trigger indirectly `solara_viz._get_scenario_defaults`. Here , we try to reconstruct the Scenario instance partly based on `model_parameters`.  This dict is populated based on the UI sliders. And so includes the rng slider. However, `rng` is not included in `_scenario_defaults`, so it is not included when creating the scenario instance and instead is kept as part of model_parameters. 

The solution is to change how we recreate the scenario instance in `_build_model_init_kwargs`. 

for example

```python
        for key, value in model_parameters.items():
            if key == 'rng':
                scenario_kwargs['key'] = value
            elif key in scenario_defaults and key not in model_init_params:
                scenario_kwargs[key] = value
            else:
                kwargs[key] = value
```

fixes the problem. This is clearly hacky, but it shows that my diagnosis is correct. A slightly more elabore solution is to change `_get_scenario_defaults`:

```python
def _get_scenario_defaults(scenario: Scenario) -> dict[str, Any]:
    """Return defaults for a scenario instance, validating scenario type."""
    scenario_class = type(scenario)
    if not issubclass(scenario_class, Scenario):
        raise ValueError(
            "Expected model.scenario to be a subclass instance of mesa.experimental.scenarios.Scenario."
        )
    defaults = getattr(scenario, "_scenario_defaults", {})
    defaults['rng'] = None

    return defaults
```

File: tests/visualization/test_solara_viz_scenarios.py
Comment (quaquel) [inline tests/visualization/test_solara_viz_scenarios.py:1]: We are moving our tests over to pytest, so please rewrite this using pytest instead of unittest. Should be a five minute job or even faster with an llm.

File: tests/examples/test_examples_viz.py [hunk 18]
```diff
@@ -16,13 +17,73 @@
...
 )
+from mesa.examples.basic.boltzmann_wealth_model.model import BoltzmannScenario
 from mesa.visualization.components import AgentPortrayalStyle
 from mesa.visualization.components.matplotlib_components import (
     PlotMatplotlib,
     SpaceMatplotlib,
 )
 
 
+def _screenshot_first_image(page, attempts=3, wait_ms=100):
```
Comment (quaquel) [inline tests/examples/test_examples_viz.py]: What is going on with these new functions? How is this relevant to the scenario issue? If it is not relevant, please consider submitting it as a separate pr

Comment (falloficaruss) [reply tests/examples/test_examples_viz.py]: sorry these were not meant to be pushed (Might have messed up while checking out with uncommitted changes)

**Comment by quaquel** [comment]
It works now on my end. I will give the code another quick check, but it seems we are almost there.

File: tests/visualization/test_solara_viz_scenarios.py [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+
+    assert scenario.density == 0.8
+    assert scenario.vision == 10
+    assert scenario.speed == 1.0  # Still default.
+
+
+@pytest.mark.skip(
+    reason="Requires interactive Solara state handling; covered indirectly."
+)
+def test_reset_with_scenario():
```
Comment (quaquel) [inline tests/visualization/test_solara_viz_scenarios.py:128]: I understand that you skip this, but you might want to test `_build_model_init_kwargs`. It would have helped in finding the bug.

Comment (falloficaruss) [reply tests/visualization/test_solara_viz_scenarios.py:128]: Okay I'll replace the skipped test with a concrete `_build_model_init_kwargs` test

**Comment by falloficaruss** [comment]
> It works now on my end. I will give the code another quick check, but it seems we are almost there.

Thank you for the help!

**Comment by EwoutH** [comment]
Great to see the progress!

I would like to also have @Sahil-Chhoker’s take on it.

**Review by quaquel — APPROVED** [review]

**Comment by quaquel** [comment]
I have tested this locally and can confirm it works. I am merging this PR, and we'll see if #3167 also works. @Sahil-Chhoker, I would still appreciate your thoughts even post-merge.

**Comment by EwoutH** [comment]
Thanks for this effort, glad to see it merged!
